819 lfric netcdf extraction

hiker opened the pull request:
Add support for kernel extraction in LFRic, fully based on jinja. It also adds an executable example for kernel extraction. Needs #955 which adds the jinja base class.
Note, driver creation is not yet supported.

hiker commented:
Ready to go (assuming that CI finishes ok).

arporter commented on lib/extract/netcdf/lfric/kernel_data_netcdf.jinja:
```diff
@@ -0,0 +1,249 @@
+        type(field_proxy_type) :: value_proxy
+
+        value_proxy = value%get_proxy()
+        call this%ExtractNetcdfBaseType%WriteArray1dDouble(name, value_proxy%data)
```
In future, could we have a `WriteArray1d` Fortran interface and then let the compiler worry about which actual implementation to call?

hiker replied:
We already have a generic `ProvideVariable` interface, which matches to `WriteArray1dDouble` (or ``WriteArray1dReal`` or so). I just had to use it :grin:

arporter replied:
Aha!

arporter commented on lib/extract/netcdf/lfric/kernel_data_netcdf.jinja:
```diff
@@ -0,0 +1,249 @@
+        ! Provide each component of the vector as an individual 1d array.
+        ! The base class will re-allocate internal array sizes if required
+        do i = 1, size(value)
+            value_proxy = value(1)%get_proxy()
```
Shouldn't that be `value(i)`?

hiker replied:
Indeed. I've opened #991 to add testing for the PSyData libraries.

arporter commented on lib/extract/netcdf/lfric/kernel_data_netcdf.jinja:
```diff
@@ -0,0 +1,249 @@
+        character(9)           :: number
+
+        ! Provide each dimension of the vector as an individual 1d array.
+        value_proxy = value(1)%get_proxy()
```
You don't need this line?

hiker replied:
No, I don't :)

arporter reviewed (changes requested):
Great stuff! I can run the example and examine the dumped netcdf file.
Please could you (briefly) document the new example in `examples/lfric/README.md` and in the user guide.
I'm really starting to feel uneasy that we don't have tests for the (fairly extensive) functionality under `lib` but that can wait for now.

hiker commented on examples/lfric/full_example_extract/README.md:
```diff
@@ -0,0 +1,32 @@
+
+This directory contains a runnable example of kernel extraction
+with LFRic. The main code is explained in more details in the
+directory ``../full_example``/
```
It should be a "." I guess - as far as I can see we usually do not have a final '/' after a directory

hiker commented on examples/lfric/full_example_extract/README.md:
```diff
@@ -0,0 +1,32 @@
+with LFRic. The main code is explained in more details in the
+directory ``../full_example``/
+
+It applies the PSyData extraction transformations (due to #637
```
Done.

hiker commented on examples/lfric/full_example_extract/README.md:
```diff
@@ -0,0 +1,32 @@
+
+The location of the PSyData wrapper library can be set with
+the environment variable ``EXTRACT_DIR``, and the location
+of the LFRic infrastructure libraries is set using.
```
Indeed.

hiker commented on examples/lfric/full_example_extract/extract_transform.py:
```diff
@@ -0,0 +1,81 @@
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
```
Done.

hiker commented on examples/lfric/full_example_extract/extract_transform.py:
```diff
@@ -0,0 +1,81 @@
+    schedule = invoke.schedule
+
+    # Enclose everything in a extract region
+    newschedule, _ = extract.apply(schedule.children,
```
I have to admit I am not sure what the current state of that actually is. I think we agreed that it is not required atm, but did we ever agree that it will stay like this (or if this was just an accidental thing)? Anyway, fixed.

arporter replied:
Yes, I think we agreed that we would do away with it altogether but it's not high priority. (I have discovered a gotcha with the NEMO API though in that you get back the Schedule at the very root of the PSyIR which is not necessarily the scehdule you gave the transformation in the first place!)

hiker commented on lib/extract/netcdf/lfric/Makefile:
```diff
@@ -0,0 +1,76 @@
+F90 ?= gfortran
+F90FLAGS ?=
+
+# For now pick up the compile environment from an
```
I reworded this a bit - the ?= should mean that you can overwrite that Make variable from the environment.

hiker commented on lib/extract/netcdf/lfric/Makefile:
```diff
@@ -0,0 +1,76 @@
+	$(F90) $(F90FLAGS) -c $<
+
+clean:
+	rm -f *.o *.mod $(LIB_NAME) psy_data_base.f90 extract_netcdf_base.f90
```
Done.

hiker commented on lib/extract/netcdf/lfric/README.md:
```diff
@@ -0,0 +1,43 @@
+``gfortran`` and the empty string. The NetCDF helper program
+``nf-config`` is used to get the NetCDF-specific include paths.
+
+.. note::
```
Ah yes, that's rst, not md. I couldn't find a replacement of this in md (an indented block), except code block, which looks bad when rendered - so I made this a sub-sub-section, and had to move it a bit further down, to clearly indicate where the note finishes.

hiker commented on lib/extract/netcdf/lfric/README.md:
```diff
@@ -0,0 +1,43 @@
+- This library uses NetCDF to store the data, so NetCDF must
+  be available on the system. It can be downloaded from
+  https://www.unidata.ucar.edu/software/netcdf/
+- The LFRIc inftrastructure library. It needs a separate
```
The first one is indeed outdated... I missed that section :) And I've added a warning to make sure to use the same version of the modules when compiling the wrapper library and LFRic.

hiker commented on lib/extract/netcdf/lfric/README.md:
```diff
@@ -0,0 +1,43 @@
+
+The following dependencies must be available:
+- This library uses NetCDF to store the data, so NetCDF must
+  be available on the system. It can be downloaded from
```
Not entirely sure what you meant with dev package, but I've added some comments about adding gfortran and a netcdf-dev package or so. Hope that's good enough.

arporter replied:
Apologies for not being clear but that's exactly what I meant :-)

hiker commented on lib/extract/netcdf/lfric/kernel_data_netcdf.jinja:
```diff
@@ -0,0 +1,249 @@
+   and for 1- to 4- dimensional arrays. It only implements the
+   required support for LFRic fields and vector-fields. Still,
+   this subroutine has to create the generic interfaces to
+   ALL these functions (the ones fromthe base class and
```
Oops :(

hiker commented on lib/extract/netcdf/lfric/kernel_data_netcdf.jinja:
```diff
@@ -0,0 +1,249 @@
+            {{all_declares|join(", &\n"+indent) }}
+
+        !> The generic interface for providing the value of variables,
+        !! which in case of the read-only verification either computes
```
Aeh ... no, it isn't.

hiker commented on lib/extract/netcdf/lfric/kernel_data_netcdf.jinja:
```diff
@@ -0,0 +1,249 @@
+Contains
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares a double precision LFRic field.
```
I've fixed that properly, stupid of me to hardcode that, since we already have a generic interface. So I've removed all references to 'double' from the field and vector field functions.

hiker commented:
> Great stuff! I can run the example and examine the dumped netcdf file.

Hooray 👍 

> Please could you (briefly) document the new example in `examples/lfric/README.md` and in the user guide.

Done. I've also opened #990 to update the naming scheme (my apologies, I never expected this to be merged on master, and just kept on using full_example_something directories. I'll clean this up after the workshop.

> I'm really starting to feel uneasy that we don't have tests for the (fairly extensive) functionality under `lib` but that can wait for now.

I _so_ agree with that. I've been quite scared for some time about this (see the `value(1)` instead of `value(i)`). See #991.

Ready for next review.

hiker commented:
Interestingly enough, using the GOceanExtract I can actually produce a 'reasonable' driver (yes, gocean changes a few things for gocean code creation, which does not take any affect on lfric):
```
  MODULE main
    IMPLICIT NONE
    CONTAINS
    SUBROUTINE main_code()
      USE testkern_w0_mod, ONLY: testkern_w0_code
      USE extract_psy_data_mod, ONLY: extract_PSyDataType
      IMPLICIT NONE
      INTEGER(KIND=i_def) cell
      REAL(KIND=8), allocatable, dimension(:,:) :: field2
      REAL(KIND=8), allocatable, dimension(:,:) :: field1_post
      REAL(KIND=8), allocatable, dimension(:,:) :: field1
      TYPE(extract_PSyDataType) extract_psy_data
      CALL extract_psy_data%OpenRead("main", "update")
      CALL extract_psy_data%ReadVariable("field1", field1)
      CALL extract_psy_data%ReadVariable("field1_post", field1_post)
      CALL extract_psy_data%ReadVariable("field2", field2)
      CALL extract_psy_data%ReadVariable("nlayers", nlayers)
      CALL extract_psy_data%ReadVariable("undf_w0", undf_w0)
      !
      ! RegionStart
      DO cell=1,field1_proxy%vspace%get_ncell()
        !
        CALL testkern_w0_code(nlayers, field1_proxy%data, field2_proxy%data, ndf_w0, undf_w0, map_w0(:,cell))
      END DO
    END SUBROUTINE main_code
  END MODULE main
```
Yes, incorrect data types, and the field accesses when creating the kernel should not be there ... but you get at least an idea of what a driver does (I removed several lines to make it short, all the variables are there, they should just be normal arrays, not using %data - which is what I had changed for gocean).

arporter commented on changelog:
```diff
@@ -214,6 +214,9 @@
 	generation time if there are any orphan OpenACC or OpenMP
 	directives e.g. OpenMP loop should be within OpenMP parallel.
 
+        69) PR #970 for #955. Use Jinja to generate the netcdf extraction
```
Not sure what this is doing in this PR - perhaps git has got confused over a merge?

hiker replied:
I believe that is actually correct - it is the merge of introducing jinja extraction, which at that stage is only dl_esm_inf. This PR then only adds LFRic.

arporter reviewed (approved):
All good now. I've taken the liberty of fixing a few typos.
I also realised that, due to the different naming scheme, the `full_example*` examples weren't being exercised by Travis. I've therefore updated the various Makefiles so that at least the `transform` parts of them are now run.
A new issue has been created regarding the naming of these examples.
